storage/index/results.go

```diff
@@ -0,0 +1,144 @@
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package index
+
+import (
+	"errors"
+
+	"github.com/m3db/m3ninx/doc"
+	"github.com/m3db/m3x/ident"
+	"github.com/m3db/m3x/pool"
+)
+
+var (
+	errUnableToAddDocMissingID = errors.New("corrupt data, unable to extract id")
+)
+
+type results struct {
+	nsID       ident.ID
+	resultsMap *ResultsMap
+
+	idPool       ident.Pool
+	bytesPool    pool.CheckedBytesPool
+	tagArrayPool TagArrayPool
+
+	pool ResultsPool
+}
+
+// NewResults returns a new results object.
+func NewResults(opts Options) Results {
+	return &results{
+		resultsMap:   newResultsMap(opts.IdentifierPool()),
+		idPool:       opts.IdentifierPool(),
+		bytesPool:    opts.CheckedBytesPool(),
+		tagArrayPool: opts.TagArrayPool(),
+		pool:         opts.ResultsPool(),
+	}
+}
+
+func (r *results) Add(d doc.Document) (added bool, err error) {
+	if len(d.ID) == 0 {
+		added = false
+		return added, errUnableToAddDocMissingID
+	}
+
+	// NB: can cast the []byte -> ident.ID to avoid an alloc
+	// before we're sure we need it.
+	tsID := ident.BytesID(d.ID)
+
+	// check if it already exists in the map.
+	if r.resultsMap.Contains(tsID) {
+		added = false
+		return added, nil
+	}
+
+	// i.e. it doesn't exist in the map, so we create the tags wrapping
+	// fields prodided by the document.
+	tags := r.tags(d.Fields)
+
+	// We use Set() instead of SetUnsafe to ensure we're taking a copy of
+	// the tsID's bytes.
+	r.resultsMap.Set(tsID, tags)
+
+	added = true
+	return added, nil
+}
+
+func (r *results) tags(fields doc.Fields) []ident.Tag {
+	tags := r.tagArrayPool.Get()[:0]
+	for _, f := range fields {
+		tags = append(tags, ident.Tag{
+			Name:  r.copyBytes(f.Name),
+			Value: r.copyBytes(f.Value),
+		})
+	}
+	return tags
+}
+
+// copyBytes copies the provided bytes into an ident.ID backed by pooled types.
+func (r *results) copyBytes(b []byte) ident.ID {
+	cb := r.bytesPool.Get(len(b))
+	cb.IncRef()
+	cb.AppendAll(b)
+	id := r.idPool.BinaryID(cb)
+	// release held reference so now the only reference to the bytes is owned by `id`
+	cb.DecRef()
+	return id
+}
+
+func (r *results) Namespace() ident.ID {
+	return r.nsID
+}
+
+func (r *results) Map() *ResultsMap {
+	return r.resultsMap
+}
+
+func (r *results) Reset(nsID ident.ID) {
+	r.nsID = nsID
+
+	// reset all values from map first
+	for _, entry := range r.resultsMap.Iter() {
+		tags := entry.Value()
+		tags.Finalize()
```

Comment:
talked offline: tags is a slice so it can't hold a ref to the pool to allow finalize to put it back. re-worked the code to do the iteration/ident finalize inline, makes the flow a little easier to reason about.

Reply:
done

Reply:
ended up doing this instead, because it reads a lot cleaner:
```
	// reset all values from map first
	for _, entry := range r.resultsMap.Iter() {
		tags := entry.Value()
		tags.Finalize()
		r.tagArrayPool.Put(tags)
	}
```

`tags.Finalize()` releases all backing ident.IDs, and the `Put()` sets the slice elems to empty value and re-sizes slice to [:0] before putting back into the pool.

Reply:
done